### DISCUSSION TITLE
The docking speed of antibody-antigen docking with Snugdock is too slow.

### DISCUSSION THREAD

**Discussion by Tangbbmc** [RosettaCommons/rosetta#140] [General]
The antibody-antigen docking with Snugdock was done on a Ubuntu 22.04 LTS with a 32 CPU with the following command: snugdock.static.linuxgccrelease -dock_pert 3 8 -partners LH_A -spin -h3_filter false -mute core.io.database -loops:refine_outer_cycles 2 -loops:max_inner_cycles 20 -detect_disulf false -antibody:check_cdr_chainbreaks false -nstruct 1000 -s $prepacked_complex.pdb. I found that the docking speed is too slow. The estimated time per run is about ten days. By the way, flag "-nstruct 1000" is the recommended setting (https://docs.rosettacommons.org/docs/latest/application_documentation/antibody/snugdock). Could you tell me how to deal with this issue? Thank you!

**Comment by vmullig** [comment]
The snugdock.static.linuxgccrelease application is using just one core. I
don't know for certain whether snugdock supports MPI parallelization, but
it probably does, which means you could get a (roughly, at best) 32 fold
speedup by using the MPI build and running on all 32 cores.

**Reply by Tangbbmc** [reply]
According to your advice, I have downloaded the source file of Rosetta 3.14 and compiled the MPI version of Rosetta using the following command: python ./scons.py  -j 30 mode=release bin extras=mpi. Finally, the generated bin folder contains 456 items (see below). I am not sure whether I have successfully compiled the MPI version. Becasure these 456 items do not mention "MPI" characters in their names. Moreover, the compiling process in terminal mentioned a error "g++: warning: switch '-ffor-scope' is no longer supported". When I run Snugdock using either snugdock.default.linuxgccrelease or snugdock.linuxgccrelease, the computation speed is still very slow. Could you guide me how to solve it? Thank you!

456 items in bin folder compiling is complete:
AbinitioRelax.default.linuxgccrelease
AbinitioRelax.linuxgccrelease
analyze_base_pairing.default.linuxgccrelease
analyze_base_pairing.linuxgccrelease
AnchoredDesign.default.linuxgccrelease
AnchoredDesign.linuxgccrelease
AnchoredPDBCreator.default.linuxgccrelease
AnchoredPDBCreator.linuxgccrelease
AnchorFinder.default.linuxgccrelease
AnchorFinder.linuxgccrelease
antibody_CR_paratope.default.linuxgccrelease
antibody_CR_paratope.linuxgccrelease
antibody.default.linuxgccrelease
antibody_designer.default.linuxgccrelease
antibody_designer.linuxgccrelease
antibody_graft.default.linuxgccrelease
antibody_graft.linuxgccrelease
antibody_H3.default.linuxgccrelease
antibody_H3.linuxgccrelease
antibody_legacy.default.linuxgccrelease
antibody_legacy.linuxgccrelease
antibody.linuxgccrelease
antibody_numbering_converter.default.linuxgccrelease
antibody_numbering_converter.linuxgccrelease
atom_energy_breakdown.default.linuxgccrelease
atom_energy_breakdown.linuxgccrelease
avrg_silent.default.linuxgccrelease
avrg_silent.linuxgccrelease
backrub.default.linuxgccrelease
backrub.linuxgccrelease
backside_hbond_finder.default.linuxgccrelease
backside_hbond_finder.linuxgccrelease
batch_distances.default.linuxgccrelease
batch_distances.linuxgccrelease
beta_peptide_modeling.default.linuxgccrelease
beta_peptide_modeling.linuxgccrelease
build_full_model.default.linuxgccrelease
build_full_model.linuxgccrelease
BuildPeptide.default.linuxgccrelease
BuildPeptide.linuxgccrelease
calc_nmr_tensor.default.linuxgccrelease
calc_nmr_tensor.linuxgccrelease
calculate_sasa.default.linuxgccrelease
calculate_sasa.linuxgccrelease
calibur.default.linuxgccrelease
calibur.linuxgccrelease
cartesian_ddg.default.linuxgccrelease
cartesian_ddg.linuxgccrelease
ca_to_allatom.default.linuxgccrelease
ca_to_allatom.linuxgccrelease
cl_complex_rescore.default.linuxgccrelease
cl_complex_rescore.linuxgccrelease
cluster_alns.default.linuxgccrelease
cluster_alns.linuxgccrelease
cluster.default.linuxgccrelease
cluster.linuxgccrelease
combine_silent.default.linuxgccrelease
combine_silent.linuxgccrelease
constel.default.linuxgccrelease
constel.linuxgccrelease
contactMap.default.linuxgccrelease
contactMap.linuxgccrelease
count_cycpep_sequences.default.linuxgccrelease
count_cycpep_sequences.linuxgccrelease
coupled_moves.default.linuxgccrelease
coupled_moves.linuxgccrelease
create_clash-based_repack_shell.default.linuxgccrelease
create_clash-based_repack_shell.linuxgccrelease
cs_rosetta_rna.default.linuxgccrelease
cs_rosetta_rna.linuxgccrelease
CstfileToTheozymePDB.default.linuxgccrelease
CstfileToTheozymePDB.linuxgccrelease
cst_info.default.linuxgccrelease
cst_info.linuxgccrelease
DARC.default.linuxgccrelease
DARC.linuxgccrelease
ddg_benchmark.default.linuxgccrelease
ddg_benchmark.linuxgccrelease
ddg_monomer.default.linuxgccrelease
ddg_monomer.linuxgccrelease
denovo_density.default.linuxgccrelease
denovo_density.linuxgccrelease
density_tools.default.linuxgccrelease
density_tools.linuxgccrelease
design_contrast_and_statistic.default.linuxgccrelease
design_contrast_and_statistic.linuxgccrelease
dgdp.default.linuxgccrelease
dgdp.linuxgccrelease
DNA_methylation.default.linuxgccrelease
DNA_methylation.linuxgccrelease
dock_glycans.default.linuxgccrelease
dock_glycans.linuxgccrelease
docking_prepack_protocol.default.linuxgccrelease
docking_prepack_protocol.linuxgccrelease
docking_protocol.default.linuxgccrelease
docking_protocol.linuxgccrelease
doug_dock_design_min_mod2_cal_cal.default.linuxgccrelease
doug_dock_design_min_mod2_cal_cal.linuxgccrelease
drrafter_error_estimation.default.linuxgccrelease
drrafter_error_estimation.linuxgccrelease
dump_capped_residue.default.linuxgccrelease
dump_capped_residue.linuxgccrelease
edge_file_generator.default.linuxgccrelease
edge_file_generator.linuxgccrelease
energy_based_clustering.default.linuxgccrelease
energy_based_clustering.linuxgccrelease
ensemble_analysis.default.linuxgccrelease
ensemble_analysis.linuxgccrelease
ensemble_generator_score12_sidechain_ver2.default.linuxgccrelease
ensemble_generator_score12_sidechain_ver2.linuxgccrelease
enzyme_design.default.linuxgccrelease
enzyme_design.linuxgccrelease
erraser2.default.linuxgccrelease
erraser2.linuxgccrelease
erraser_minimizer.default.linuxgccrelease
erraser_minimizer.linuxgccrelease
exposed_strand_finder.default.linuxgccrelease
exposed_strand_finder.linuxgccrelease
extract_atomtree_diffs.default.linuxgccrelease
extract_atomtree_diffs.linuxgccrelease
extract_motifs.default.linuxgccrelease
extract_motifs.linuxgccrelease
extract_pdbs.default.linuxgccrelease
extract_pdbs.linuxgccrelease
fast_clustering.default.linuxgccrelease
fast_clustering.linuxgccrelease
FiberDiffractionFreeSet.default.linuxgccrelease
FiberDiffractionFreeSet.linuxgccrelease
fit_helix_in_map.default.linuxgccrelease
fit_helix_in_map.linuxgccrelease
fix_alignment_to_match_pdb.default.linuxgccrelease
fix_alignment_to_match_pdb.linuxgccrelease
fixbb.default.linuxgccrelease
fixbb.linuxgccrelease
FlexPepDocking.default.linuxgccrelease
FlexPepDocking.linuxgccrelease
FloppyTail.default.linuxgccrelease
FloppyTail.linuxgccrelease
fragment_picker.default.linuxgccrelease
fragment_picker.linuxgccrelease
full_length_model.default.linuxgccrelease
full_length_model.linuxgccrelease
gen_apo_grids.default.linuxgccrelease
gen_apo_grids.linuxgccrelease
generate_database_cache.default.linuxgccrelease
generate_database_cache.linuxgccrelease
generate_starting_zinc_sites.default.linuxgccrelease
generate_starting_zinc_sites.linuxgccrelease
gen_lig_grids.default.linuxgccrelease
gen_lig_grids.linuxgccrelease
glycan_clash_check.default.linuxgccrelease
glycan_clash_check.linuxgccrelease
GlycanDock.default.linuxgccrelease
GlycanDock.linuxgccrelease
glycan_info.default.linuxgccrelease
glycan_info.linuxgccrelease
glycosyltransfer.default.linuxgccrelease
glycosyltransfer.linuxgccrelease
grower_prep.default.linuxgccrelease
grower_prep.linuxgccrelease
hbs_design.default.linuxgccrelease
hbs_design.linuxgccrelease
HDXEnergy.default.linuxgccrelease
HDXEnergy.linuxgccrelease
helical_bundle_predict.default.linuxgccrelease
helical_bundle_predict.linuxgccrelease
helix_from_sequence.default.linuxgccrelease
helix_from_sequence.linuxgccrelease
homodimer_design.default.linuxgccrelease
homodimer_design.linuxgccrelease
homodimer_maker.default.linuxgccrelease
homodimer_maker.linuxgccrelease
hydrate.default.linuxgccrelease
hydrate.linuxgccrelease
idealize_jd2.default.linuxgccrelease
idealize_jd2.linuxgccrelease
identify_cdr_clusters.default.linuxgccrelease
identify_cdr_clusters.linuxgccrelease
identify_ligand_motifs.default.linuxgccrelease
identify_ligand_motifs.linuxgccrelease
incorporate_motifs.default.linuxgccrelease
incorporate_motifs.linuxgccrelease
InterfaceAnalyzer.default.linuxgccrelease
InterfaceAnalyzer.linuxgccrelease
interface_energy.default.linuxgccrelease
interface_energy.linuxgccrelease
iterhybrid_selector.default.linuxgccrelease
iterhybrid_selector.linuxgccrelease
ligand_discovery_search_protocol.default.linuxgccrelease
ligand_discovery_search_protocol.linuxgccrelease
ligand_dock.default.linuxgccrelease
ligand_dock.linuxgccrelease
ligand_rpkmin.default.linuxgccrelease
ligand_rpkmin.linuxgccrelease
ligand_transform_with_pcs.default.linuxgccrelease
ligand_transform_with_pcs.linuxgccrelease
loodo.default.linuxgccrelease
loodo.linuxgccrelease
loopmodel.default.linuxgccrelease
loopmodel.linuxgccrelease
make_exemplar.default.linuxgccrelease
make_exemplar.linuxgccrelease
make_motif_tables.default.linuxgccrelease
make_motif_tables.linuxgccrelease
make_ray_files.default.linuxgccrelease
make_ray_files.linuxgccrelease
MakeRotLib.default.linuxgccrelease
MakeRotLib.linuxgccrelease
MakeSegmentFile.default.linuxgccrelease
MakeSegmentFile.linuxgccrelease
match.default.linuxgccrelease
match.linuxgccrelease
membrane_abinitio2.default.linuxgccrelease
membrane_abinitio2.linuxgccrelease
metal_site_finder.default.linuxgccrelease
metal_site_finder.linuxgccrelease
mg_modeler.default.linuxgccrelease
mg_modeler.linuxgccrelease
minimize.default.linuxgccrelease
minimize.linuxgccrelease
minimize_ppi.default.linuxgccrelease
minimize_ppi.linuxgccrelease
minimize_with_cst.default.linuxgccrelease
minimize_with_cst.linuxgccrelease
minirosetta.default.linuxgccrelease
minirosetta_graphics.default.linuxgccrelease
minirosetta_graphics.linuxgccrelease
minirosetta.linuxgccrelease
mp_dock.default.linuxgccrelease
mp_dock.linuxgccrelease
mp_dock_setup.default.linuxgccrelease
mp_dock_setup.linuxgccrelease
mp_domain_assembly.default.linuxgccrelease
mp_domain_assembly.linuxgccrelease
mpi_msd.default.linuxgccrelease
mpi_msd.linuxgccrelease
mp_lipid_acc.default.linuxgccrelease
mp_lipid_acc.linuxgccrelease
mp_mutate_relax.default.linuxgccrelease
mp_mutate_relax.linuxgccrelease
mp_range_relax.default.linuxgccrelease
mp_range_relax.linuxgccrelease
mp_seqrecov.default.linuxgccrelease
mp_seqrecov.linuxgccrelease
mp_span_from_pdb.default.linuxgccrelease
mp_span_from_pdb.linuxgccrelease
mp_symdock.default.linuxgccrelease
mp_symdock.linuxgccrelease
mp_transform.default.linuxgccrelease
mp_transform.linuxgccrelease
mp_viewer.default.linuxgccrelease
mp_viewer.linuxgccrelease
mr_protocols.default.linuxgccrelease
mr_protocols.linuxgccrelease
multistage_rosetta_scripts.default.linuxgccrelease
multistage_rosetta_scripts.linuxgccrelease
mutate.default.linuxgccrelease
mutate.linuxgccrelease
N-terminal_acetylation.default.linuxgccrelease
N-terminal_acetylation.linuxgccrelease
nucleobase_sample_around.default.linuxgccrelease
nucleobase_sample_around.linuxgccrelease
oop_design.default.linuxgccrelease
oop_design.linuxgccrelease
optE_parallel.default.linuxgccrelease
optE_parallel.linuxgccrelease
packing_angle.default.linuxgccrelease
packing_angle.linuxgccrelease
packstat.default.linuxgccrelease
packstat.linuxgccrelease
parcs_ccs_calc.default.linuxgccrelease
parcs_ccs_calc.linuxgccrelease
parse_rosetta_script.default.linuxgccrelease
parse_rosetta_script.linuxgccrelease
partial_thread.default.linuxgccrelease
partial_thread.linuxgccrelease
pepspec_anchor_dock.default.linuxgccrelease
pepspec_anchor_dock.linuxgccrelease
pepspec.default.linuxgccrelease
pepspec.linuxgccrelease
PeptideDeriver.default.linuxgccrelease
PeptideDeriver.linuxgccrelease
peptoid_design.default.linuxgccrelease
peptoid_design.linuxgccrelease
performance_benchmark.default.linuxgccrelease
performance_benchmark.linuxgccrelease
per_residue_energies.default.linuxgccrelease
per_residue_energies.linuxgccrelease
per_residue_solvent_exposure.default.linuxgccrelease
per_residue_solvent_exposure.linuxgccrelease
phosphorylation.default.linuxgccrelease
phosphorylation.linuxgccrelease
pH_protocol.default.linuxgccrelease
pH_protocol.linuxgccrelease
pmut_scan_parallel.default.linuxgccrelease
pmut_scan_parallel.linuxgccrelease
pocket_measure.default.linuxgccrelease
pocket_measure.linuxgccrelease
pocket_relax.default.linuxgccrelease
pocket_relax.linuxgccrelease
pocket_suggest_target_residues_by_ddg.default.linuxgccrelease
pocket_suggest_target_residues_by_ddg.linuxgccrelease
pose_from_saccharide_sequence.default.linuxgccrelease
pose_from_saccharide_sequence.linuxgccrelease
recces.default.linuxgccrelease
recces.linuxgccrelease
recces_turner.default.linuxgccrelease
recces_turner.linuxgccrelease
recon.default.linuxgccrelease
recon.linuxgccrelease
relax.default.linuxgccrelease
relax.linuxgccrelease
remodel.default.linuxgccrelease
remodel.linuxgccrelease
report_hbonds_for_plugin.default.linuxgccrelease
report_hbonds_for_plugin.linuxgccrelease
resample_full_model.default.linuxgccrelease
resample_full_model.linuxgccrelease
ResidueDisorder.default.linuxgccrelease
ResidueDisorder.linuxgccrelease
residue_energy_breakdown.default.linuxgccrelease
residue_energy_breakdown.linuxgccrelease
restype_converter.default.linuxgccrelease
restype_converter.linuxgccrelease
revert_design_to_native.default.linuxgccrelease
revert_design_to_native.linuxgccrelease
rna_cluster.default.linuxgccrelease
rna_cluster.linuxgccrelease
rna_database.default.linuxgccrelease
rna_database.linuxgccrelease
rna_denovo.default.linuxgccrelease
rna_denovo.linuxgccrelease
rna_design.default.linuxgccrelease
rna_design.linuxgccrelease
rna_extract.default.linuxgccrelease
rna_extract.linuxgccrelease
rna_features.default.linuxgccrelease
rna_features.linuxgccrelease
rna_graft.default.linuxgccrelease
rna_graft.linuxgccrelease
rna_helix.default.linuxgccrelease
rna_helix.linuxgccrelease
rna_idealize.default.linuxgccrelease
rna_idealize.linuxgccrelease
rna_minimize.default.linuxgccrelease
rna_minimize.linuxgccrelease
rna_motif.default.linuxgccrelease
rna_motif.linuxgccrelease
rna_predict_chem_map.default.linuxgccrelease
rna_predict_chem_map.linuxgccrelease
rna_score.default.linuxgccrelease
rna_score.linuxgccrelease
rna_suitename.default.linuxgccrelease
rna_suitename.linuxgccrelease
rna_thread_and_minimize.default.linuxgccrelease
rna_thread_and_minimize.linuxgccrelease
rna_thread.default.linuxgccrelease
rna_thread.linuxgccrelease
r_noe_assign.default.linuxgccrelease
r_noe_assign.linuxgccrelease
rnp_ddg.default.linuxgccrelease
rnp_ddg.linuxgccrelease
rosettaDNA.default.linuxgccrelease
rosettaDNA.linuxgccrelease
rosetta_scripts.default.linuxgccrelease
rosetta_scripts_jd3.default.linuxgccrelease
rosetta_scripts_jd3.linuxgccrelease
rosetta_scripts.linuxgccrelease
rotamer_recovery.default.linuxgccrelease
rotamer_recovery.linuxgccrelease
sc.default.linuxgccrelease
sc.linuxgccrelease
score_aln2.default.linuxgccrelease
score_aln2.linuxgccrelease
score_aln.default.linuxgccrelease
score_aln.linuxgccrelease
score.default.linuxgccrelease
score_jd2.default.linuxgccrelease
score_jd2.linuxgccrelease
score.linuxgccrelease
segment_file_generator.default.linuxgccrelease
segment_file_generator.linuxgccrelease
select_best_unique_ligand_poses.default.linuxgccrelease
select_best_unique_ligand_poses.linuxgccrelease
sequence_recovery.default.linuxgccrelease
sequence_recovery.linuxgccrelease
sequence_tolerance.default.linuxgccrelease
sequence_tolerance.linuxgccrelease
sewing_hasher.default.linuxgccrelease
sewing_hasher.linuxgccrelease
shobuns.default.linuxgccrelease
shobuns.linuxgccrelease
SID_ERMS_prediction.default.linuxgccrelease
SID_ERMS_prediction.linuxgccrelease
simple_cycpep_predict.default.linuxgccrelease
simple_cycpep_predict.linuxgccrelease
snugdock.default.linuxgccrelease
snugdock.linuxgccrelease
spheregrinder.default.linuxgccrelease
spheregrinder.linuxgccrelease
stepwise.default.linuxgccrelease
stepwise.linuxgccrelease
struc_set_fragment_picker.default.linuxgccrelease
struc_set_fragment_picker.linuxgccrelease
super_aln.default.linuxgccrelease
super_aln.linuxgccrelease
supercharge.default.linuxgccrelease
supercharge.linuxgccrelease
surface_docking.default.linuxgccrelease
surface_docking.linuxgccrelease
swa_protein_main.default.linuxgccrelease
swa_protein_main.linuxgccrelease
swa_rna_main.default.linuxgccrelease
swa_rna_main.linuxgccrelease
swa_rna_util.default.linuxgccrelease
swa_rna_util.linuxgccrelease
SymDock.default.linuxgccrelease
SymDock.linuxgccrelease
tcrmodel.default.linuxgccrelease
tcrmodel.linuxgccrelease
template_features.default.linuxgccrelease
template_features.linuxgccrelease
thermal_sampler.default.linuxgccrelease
thermal_sampler.linuxgccrelease
theta_ligand.default.linuxgccrelease
theta_ligand.linuxgccrelease
torsional_potential_corrections.default.linuxgccrelease
torsional_potential_corrections.linuxgccrelease
transform_loodo.default.linuxgccrelease
transform_loodo.linuxgccrelease
trRosetta.default.linuxgccrelease
trRosetta.linuxgccrelease
UBQ_E2_thioester.default.linuxgccrelease
UBQ_E2_thioester.linuxgccrelease
UBQ_Gp_CYD-CYD.default.linuxgccrelease
UBQ_Gp_CYD-CYD.linuxgccrelease
UBQ_Gp_LYX-Cterm.default.linuxgccrelease
UBQ_Gp_LYX-Cterm.linuxgccrelease
UnfoldedStateEnergyCalculator.default.linuxgccrelease
UnfoldedStateEnergyCalculator.linuxgccrelease
validate_database.default.linuxgccrelease
validate_database.linuxgccrelease
validate_rosetta_script.default.linuxgccrelease
validate_rosetta_script.linuxgccrelease
validate_silent.default.linuxgccrelease
validate_silent.linuxgccrelease
vip.default.linuxgccrelease
vip.linuxgccrelease
zinc2_homodimer_design.default.linuxgccrelease
zinc2_homodimer_design.linuxgccrelease
zinc2_homodimer_setup.default.linuxgccrelease
zinc2_homodimer_setup.linuxgccrelease
zinc_heterodimer_design.default.linuxgccrelease
zinc_heterodimer_design.linuxgccrelease
zinc_statistic_generator.default.linuxgccrelease

**Comment by vmullig** [comment]
> According to your advice, I have downloaded the source file of Rosetta 3.14 and compiled the MPI version of Rosetta using the following command: python ./scons.py -j 30 mode=release bin extras=mpi. Finally, the generated bin folder contains 456 items (see below). I am not sure whether I have successfully compiled the MPI version. Becasure these 456 items do not mention "MPI" characters in their names. Moreover, the compiling process in terminal mentioned a error "g++: warning: switch '-ffor-scope' is no longer supported". When I run Snugdock using either snugdock.default.linuxgccrelease or snugdock.linuxgccrelease, the computation speed is still very slow. Could you guide me how to solve it? Thank you!

Hmm.  Looks like something's going wrong with the MPI compilation, if you're not getting files of the form `<some_name>.mpi.linuxgccrelease` produced.  It's not the `ffor-scope` warning -- that's just a known thing with newer versions of GCC that doesn't prevent compilation.  Check the compliation log, though, for errors; it is likely failing.  Full compilation in MPI mode will likely take ~10 core-hours on a modern CPU (i.e. about 20 minutes if split over 30 cores with -j30), so if it took far less than that, it likely failed.

**Comment by Tangbbmc** [comment]
Thank you for your reply! Where is the compliation log? I have uploaded the final running log shown in terminal.
![IMG_20240905_095805](https://github.com/user-attachments/assets/eba65b2c-8ccc-415b-a051-9e893117a604)

**Reply by vmullig** [reply]
> Thank you for your reply! Where is the compliation log? I have uploaded the final running log shown in terminal.

Oh, weird!  It looks like it's just doing a default compilation.

Hmm.  Try putting "bin" last in your compilation command:

```
./scons.py -j 30 mode=release extras=mpi bin
```

**Reply by vmullig** [reply]
(The individual compilation commands should invoke the MPI C++ compiler -- usually `mpicxx` or `mpic++` or some such.)

**Comment by Tangbbmc** [comment]
Your advice is excellent. I have successfully installed the mpi version of rosetta. I really appreciated your help!

发件人："Vikram K. Mulligan" ***@***.***>
发送日期：2024-09-05 13:54:56
收件人：RosettaCommons/rosetta ***@***.***>
抄送人：Wanggang ***@***.***>,Author ***@***.***>
主题：Re: [RosettaCommons/rosetta] The docking speed of antibody-antigen docking with Snugdock is too slow. (Discussion #140)

Thank you for your reply! Where is the compliation log? I have uploaded the final running log shown in terminal.
Oh, weird! It looks like it's just doing a default compilation.
Hmm. Try putting "bin" last in your compilation command:
./scons.py -j 30 mode=release extras=mpi bin

**Comment by Tangbbmc** [comment]
Ok. I got it. Thank you very much!

发件人："Vikram K. Mulligan" ***@***.***>
发送日期：2024-09-05 13:56:00
收件人：RosettaCommons/rosetta ***@***.***>
抄送人：Wanggang ***@***.***>,Author ***@***.***>
主题：Re: [RosettaCommons/rosetta] The docking speed of antibody-antigen docking with Snugdock is too slow. (Discussion #140)

(The individual compilation commands should invoke the MPI C++ compiler -- usually mpicxx or mpic++ or some such.)

**Comment by vmullig** [comment]
>Ok. I got it. Thank you very much!

OK, great!  The next step is running it.  I double-checked the source code, and the `snugdock` application does indeed use JD2 (version 2 of the Rosetta job distributor), which supports MPI.  To run on multiple cores on a single node, you'll need to check what your MPI run command is on your system.  On most systems (for instance, if you're using OpenMPI), it's `mpirun`; on some, it's `aprun` or `mpiexec`.  Assuming it's `mpirun`, you run an MPI program on a single node like this:

```
mpirun -np <# of processes to launch> <Rosetta executable name>.mpi.linuxgccrelease <Rosetta flags here, or @flags file> >out.log 2>err.log
```

If this is still too slow, and if you have access to a large computing cluster, you can also run on many cores on many nodes across the cluster.  You'll have to refer to your cluster's documentation for the syntax, however.

A few notes:
- In MPI mode, it's best to use the `-mute all` flag to silence most output, and then to turn on the output from only those modules that you need with the `-unmute <module name>` flag.  This is because i/o can be quite costly, if you've got a lot of processes all contending to write to the same data stream -- it can actually mean that most of your processes end up idling while waiting for access, rather than doing the actual computation.
- In MPI mode, there's a Rosetta option to write output from each process to a different output logfile, which can be handy.  It's `-mpi_tracer_to_file <logfile base name>`.  For instance, `-mpi_tracer_to_file output.txt` will produce a series of logfiles, `output0.txt`, `output1.txt`, `output2.txt`, etc., one for each MPI process.  It is _still_ important to mute any unnecessary output, though: contention for disk write can still slow down your application's execution with a lot of processes writing output.

**Reply by vmullig** [reply]
One last thing to be aware of: many flavours of MPI do not treat hardware threads as separate cores.  Many modern CPUs advertise, for example, 64 threads, but have only 32 physical cores, relying on hyper-threading to allow performance _somewhat_ close to what you'd get with 64 cores for _some_ tasks.  Be sure to either (a) launch as many processes as you have _physical cores_, or (b) use the `-oversubscribe` flag after your `mpirun` command to allow the launch of more processes than you have physical cores.  (Under most circumstances, hyper-threading does _not_ give you great performance, especially with MPI processes that could be doing totally different things at any given time.)

**Reply by Tangbbmc** [reply]
OK, thank you very much for your further advice and guidance. I have tested mpirun command (at least 20 physical cores available for my 32 CPU system). The computation speed has increased a lot.  Here, I would like to ask an question regarding result evaluation. The resulting score.sc file contained many evaluation parameters for each conformation of the docked antiboby-antigen complex. Then, how to evaluate the binding affinity between antibody and antigen using the score.sc file?

**Reply by everyday847** [reply]
If the scorefile has a column like `dG_separated`, that is the parameter that describes the binding interaction the best. I would only really look at the `dG_separated` of complexes with good overall energies -- maybe take the top 5% by score, then rank by `dG_separated`.

That said, empirically, this value is only weakly correlated with binding affinity (i.e., very good values are somewhat more likely to be strong binders than very bad values). Generalizable protein-protein binding affinity prediction across different antigens (or, indeed, across different antibody germlines) remains unsolved.

**Reply by Tangbbmc** [reply]
@everyday847 Ok. Thank you for your kind advice!

**Comment by roccomoretti** [comment]
Marking this as "closed", as I hope your question has been answered. If that assessment is incorrect, feel free to re-open or open up a new discussion with more details.
